### PR TITLE
Cldspt 9295 wmic performance

### LINKED ISSUE #27: Command does not work
Command in control V-63873 and any control that has   
get_domain_sid = command('wmic useraccount get sid | FINDSTR /V SID | Select -First 2').stdout.strip
 domain_sid = get_domain_sid[9..40]

does not work, the first problem is that First 2 would only look at the first two local accounts, the second problem is that if you modify this command to get_domain_sid = command('wmic group get Name,SID | FINDSTR /C:"Domain Users"').stdout.strip. you will get the right SID but Inspec does not read domain objects with command or powershell it only sees local accounts

### LINKED ISSUE #31: V-73775.rb Skip Wording is confusing
[controls/V-73775.rb] (https://github.com/mitre/microsoft-windows-server-2016-stig-baseline/blob/master/controls/V-73775.rb)

Reads ```This system is a domain controller, therefore this control is not applicable as it only applies to domain controllers```

### PR DESCRIPTION
**Pull request by gregbeck-cms**
The Windows CMD 'wmic useraccount get sid' in these several controls would not complete at all, or would exit with an error message on my domain-joined Win 2016 instances. The result was that the InSpec scan would not complete before its AWS SSM document's timeout of 3600 seconds. Per https://stackoverflow.com/a/55387641 this wmic command would query across _all_ users in a domain, kinda like doing a full table scan in an RDBMS, not performant except in the tiniest of domains. The result of the wmic command timing out was that the _loading_ of the control, _not its actual execution_ during a compliance scan would time out.

The result was that the scan would hang forever loading just one of these controls, e.g. from `inspec exec . -l debug` that never progressed beyond loading V-73759:

`[2021-07-02T21:33:25+00:00] DEBUG: Loading ./controls/V-73759.rb into #<Inspec::ProfileContext:0x000000000b4fb4c8>`

@ejaronne suggested I use https://github.com/mitre/microsoft-windows-server-2019-stig-baseline/blob/69ebfebfa409ff35f35e90b30a4b1bbc8b956180/controls/V-93009.rb#L103-L114 (thanks https://github.com/burnsjared0415!) to avoid issues with a _local_ Administrator using 'whoami' on a domain-joined OS, the local user SID substrings wouldn't match the domain SID. Here's an example of the changed controls scan results on a domain-joined, non-domain-contoller test instance that did not have a compliant configuration:

  [FAIL]  V-73759: The Deny access to this computer from the network user right on member
    servers must be configured to prevent access from highly privileged domain
    accounts and local accounts on domain systems, and from unauthenticated access
    on all systems. (2 failed)
     [FAIL]  Security Policy SeDenyNetworkLogonRight is expected to include "S-1-5-21-_my domain SID redacted_-512
"
     expected ["S-1-5-32-546"] to include "S-1-5-21-_my domain SID redacted_-512"
     [FAIL]  Security Policy SeDenyNetworkLogonRight is expected to include "S-1-5-21-_my domain SID redacted_-519
"
     expected ["S-1-5-32-546"] to include "S-1-5-21-_my domain SID redacted_-519"

  [FAIL]  V-73763: The Deny log on as a batch job user right on member servers must be
    configured to prevent access from highly privileged domain accounts on domain
    systems and from unauthenticated access on all systems. (2 failed)
     [FAIL]  Security Policy SeDenyNetworkLogonRight is expected to include "S-1-5-21-_my domain SID redacted_-512
"
     expected ["S-1-5-32-546"] to include "S-1-5-21-_my domain SID redacted_-512"
     [FAIL]  Security Policy SeDenyNetworkLogonRight is expected to include "S-1-5-21-_my domain SID redacted_-519
"
     expected ["S-1-5-32-546"] to include "S-1-5-21-_my domain SID redacted_-519"

  [FAIL]  V-73767: The Deny log on as a service user right on member servers must be
    configured to prevent access from highly privileged domain accounts on domain
    systems. No other groups or accounts must be assigned this right. (2 failed)
     [FAIL]  Security Policy SeDenyNetworkLogonRight is expected to include "S-1-5-21-_my domain SID redacted_-512
"
     expected ["S-1-5-32-546"] to include "S-1-5-21-_my domain SID redacted_-512"
     [FAIL]  Security Policy SeDenyNetworkLogonRight is expected to include "S-1-5-21-_my domain SID redacted_-519
"
     expected ["S-1-5-32-546"] to include "S-1-5-21-_my domain SID redacted_-519"

  [FAIL]  V-73771: The Deny log on locally user right on member servers must be
    configured to prevent access from highly privileged domain accounts on domain
    systems and from unauthenticated access on all systems. (2 failed)
     [FAIL]  Security Policy SeDenyNetworkLogonRight is expected to include "S-1-5-21-_my domain SID redacted_-512
"
     expected ["S-1-5-32-546"] to include "S-1-5-21-_my domain SID redacted_-512"
     [FAIL]  Security Policy SeDenyNetworkLogonRight is expected to include "S-1-5-21-_my domain SID redacted_-519
"
     expected ["S-1-5-32-546"] to include "S-1-5-21-_my domain SID redacted_-519"

 [FAIL]  V-73775: The Deny log on through Remote Desktop Services user right on member
   servers must be configured to prevent access from highly privileged domain
   accounts and all local accounts on domain systems and from unauthenticated
   access on all systems. (3 failed)
    [FAIL]  Security Policy SeDenyNetworkLogonRight is expected to include "S-1-5-21-_my domain SID redacted_-512
"
    expected ["S-1-5-32-546"] to include "S-1-5-21-_my domain SID redacted_-512"
    [FAIL]  Security Policy SeDenyNetworkLogonRight is expected to include "S-1-5-21-_my domain SID redacted_-519
    expected ["S-1-5-32-546"] to include "S-1-5-21-_my domain SID redacted_-519"
    [FAIL]  Security Policy SeDenyRemoteInteractiveLogonRight is expected to include "S-1-2-0"
    expected ["S-1-5-21-_my SID redacted_", "S-1-5-21-_my SID redacted_", "S-1-5-21-_my SID redacted_", "S-1-5-21-_my SID redacted_", "S-1-5-32-546"] to include "S-1-2-0"

The entire scan now completes in less than 2 minutes. I tested these changes and got expected results on one of our domain-joined gold image instances and some Windows 2016 AWS AMIs not domain-joined:
Windows_Server-2016-English-Full-Base-2021.06.09 ami-0d39d14f5f1a82d2b
Windows_Server-2016-English-STIG-Full-2021.06.09 ami-0093f2bab85e0c9d9

**_Please test this and ruthlessly poke holes before merging_**, my ignorance of Windows is surpassed only by my ignorance of Ruby/InSpec.

Many thanks for this profile, makes my life a lot easier!

Greg

Fixes #27 
Fixes #31

### CODE REVIEW

**Comment by gregbeck-cms** [comment]
There is a simpler built-in way to get the domain SID that does not require enabling the ActiveDirectory-Powershell feature, I'll make the change, test, and push to this branch.

**Comment by ejaronne** [comment]
@burnsjared0415 , could the technique used at https://github.com/mitre/microsoft-windows-server-2019-stig-baseline/blob/69ebfebfa409ff35f35e90b30a4b1bbc8b956180/controls/V-93009.rb#L103-L114 be used effectively on this 2016 baseline?

**Comment by gregbeck-cms** [comment]
Hi @ejaronne and @aaronlippold , this code works, but I am confused about why I need so many 'end's at the, well, end of the files. It seems like there are too many in each file, I couldn't get the indentation to work out nicely, may be a non-issue.

**Review by ejaronne — CHANGES_REQUESTED** [review]
I'm concerned about maintaining the purity of the original intent/content from the STIG author, especially **check and fix text**. Any deviation (however well meaning) from what a user sees in the original STIG guidance could cause confusion distrust as to the scope of the checks. Recommend instead, for example wrt V-73759.rb:
Leave the original fix text as is, then append:
"[The following text is not part of the original STIG, but added as helpful context from the InSpec profile test code developer (@gregbeck-cms)]:"
and add your statements, such as:
"  Domain Systems Only:
  - Enterprise Admins group (SID* S-1-5-21-root domain-519)
  - Domain Admins group (SID* S-1-5-21-domain-512)
  - Local account (SID* S-1-2-0) and member of Administrators group (SID S-1-5-32-544) or Local account
  (see Note below)
  All Systems:
  - Guests group (SID* S-1-5-32-546)
  * See SIDs in https://docs.microsoft.com/en-us/troubleshoot/windows-server/identity/security-identifiers-in-windows"

**Comment by gregbeck-cms** [comment]
I have removed the deviations from the STIG text.

**Review by ejaronne — APPROVED** [review]

File: controls/V-73763.rb [hunk 1]
```diff
@@ -48,11 +49,12 @@
...
 
   Domain Systems Only:
-  - Enterprise Admins Group
-  - Domain Admins Group
+  - Enterprise Admins group 
+  - Domain Admins group 
 
   All Systems:
-  - Guests Group"
+  - Guests group
```
Comment (karikarshivani) [inline controls/V-73763.rb:56]: Missing the end double-quotes - which seems to be leading to the following code to be treated as a string

File: controls/V-73767.rb [hunk 2]
```diff
@@ -48,8 +48,9 @@
   Deny log on as a service to include the following:
 
   Domain systems:
-  - Enterprise Admins Group
-  - Domain Admins Group"
+  - Enterprise Admins group 
+  - Domain Admins group
```
Comment (karikarshivani) [inline controls/V-73767.rb:52]: Missing the end double-quotes

File: controls/V-73771.rb [hunk 3]
```diff
@@ -51,15 +51,17 @@
...
-  - Domain Admins Group
+  - Enterprise Admins group 
+  - Domain Admins group 
 
   Systems dedicated to the management of Active Directory (AD admin platforms,
   see V-36436 in the Active Directory Domain STIG) are exempt from this.
 
   All Systems:
-  - Guests Group"
+  - Guests group
```
Comment (karikarshivani) [inline controls/V-73771.rb:63]: Missing the end double-quotes

File: controls/V-73759.rb [hunk 4]
```diff
@@ -100,18 +100,29 @@
...
+        enterprise_admin_sid = json(command: enterprise_admin_sid_query).params
+ 
         describe security_policy do
-          its('SeDenyNetworkLogonRight') { should include "S-1-21-#{domain_sid}-512" }
+          its('SeDenyNetworkLogonRight') { should include "#{domain_admin_sid}" }
         end
         describe security_policy do
-          its('SeDenyNetworkLogonRight') { should include "S-1-21-#{domain_sid}-519" }
+          its('SeDenyNetworkLogonRight') { should include "#{enterprise_admin_sid}" }
         end
```
[the remark this answers is not in the document]
Comment (karikarshivani) [reply controls/V-73759.rb]: Updated this control to reflect the proposal for approval

File: controls/V-73759.rb [hunk 5]
```diff
@@ -78,56 +78,54 @@
...
+    impact 0.0
+    desc 'This system is a domain controller, therefore this control is not applicable as it only applies to member servers and standalone systems'
+    describe 'This system is a domain controller, therefore this control is not applicable as it only applies to member servers and standalone systems' do
+      skip 'This system is a domain controller, therefore this control is not applicable as it only applies to member servers and standalone systems'
+    end
+  # elsif is_AD_only_system
+  #   impact 0.0
+  #   desc 'This system is dedicated to the management of Active Directory, therefore this system is exempt from this control'
+  #   describe 'This system is dedicated to the management of Active Directory, therefore this system is exempt from this control' do
+  #     skip 'This system is dedicated to the management of Active Directory, therefore this system is exempt from this control'
```
Comment (karikarshivani) [inline controls/V-73759.rb:94]: Check commented out until further research

File: controls/V-73759.rb [hunk 6]
```diff
@@ -78,56 +78,54 @@
...
+      end
+
+      describe.one do
         describe security_policy do
-          its('SeDenyNetworkLogonRight') { should include "#{domain_admin_sid}" }
+            its('SeDenyNetworkLogonRight') { should include "S-1-5-113" }
         end
         describe security_policy do
-          its('SeDenyNetworkLogonRight') { should include "#{enterprise_admin_sid}" }
+            its('SeDenyNetworkLogonRight') { should include "S-1-5-114" }
```
Comment (karikarshivani) [inline controls/V-73759.rb]: Added the SIDs for local accounts as per the updated STIG guidance. Test pending.

**Comment by karikarshivani** [comment]
[MemberServer-Fail.json.txt](https://github.com/mitre/microsoft-windows-server-2016-stig-baseline/files/6887542/MemberServer-Fail.json.txt)
[MemberServer-Pass.json.txt](https://github.com/mitre/microsoft-windows-server-2016-stig-baseline/files/6887543/MemberServer-Pass.json.txt)
[Standalone-Fail.json.txt](https://github.com/mitre/microsoft-windows-server-2016-stig-baseline/files/6887544/Standalone-Fail.json.txt)
[Standalone-Pass.json.txt](https://github.com/mitre/microsoft-windows-server-2016-stig-baseline/files/6887545/Standalone-Pass.json.txt)

File: controls/V-73759.rb (hunk 6 above)
Comment (karikarshivani) [reply controls/V-73759.rb]: Tested after adding Local accounts where needed

File: controls/V-73759.rb (hunk 5 above)
Comment (karikarshivani) [reply controls/V-73759.rb]: Checked and included code where needed

**Review by karikarshivani — APPROVED** [review]

**Pull request merged**